phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java

```diff
@@ -0,0 +1,280 @@
+                    return Optional.of(constructor);
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static List<DTOContainer<?>> findAllDTOs() throws ClassNotFoundException {
+        final List<DTOContainer<?>> fondJsonCreatorClasses = new ArrayList<>();
+
+        final ClassPathScanningCandidateComponentProvider scanner = new ClassPathScanningCandidateComponentProvider(
+                true);
+        scanner.addIncludeFilter((metadataReader, metadataReaderFactory) -> true);
+
+        final Stream<String> classNamesStream = scanner.findCandidateComponents("org.phoenicis").stream()
+                .map(BeanDefinition::getBeanClassName);
+        final Iterable<String> classNames = classNamesStream::iterator;
+
+        for (String className : classNames) {
+            final Class<?> clazz = Class.forName(className);
+            if (isJsonClass(clazz)) {
+                fondJsonCreatorClasses.add(new DTOContainer<>(clazz));
+            }
+        }
+
+        return fondJsonCreatorClasses;
+    }
+
+    private static boolean isJsonClass(Class<?> clazz) {
+        return getConstructorAnnotatedWithJsonCreator(clazz).isPresent()
+                || getConstructorHavingJsonBuilder(clazz).isPresent();
+    }
+
+    @Test
+    public void testConstructor() throws ReflectiveOperationException, URISyntaxException {
+        final Object dto = newDTOInstance(DTOContainer.getClazz());
+        assertNotNull(dto);
+    }
+
+    @Test
+    public void testHashCodes() throws ReflectiveOperationException, URISyntaxException {
+        final Object dto1 = newDTOInstance(DTOContainer.getClazz());
+        final Object dto2 = newDTOInstance(DTOContainer.getClazz());
+        assertEquals(dto1, dto2);
+        assertEquals(dto1.hashCode(), dto2.hashCode());
+    }
+
+    @Test
+    public void testEqualsTrivialCases() throws ReflectiveOperationException, URISyntaxException {
+        final Object dto = newDTOInstance(DTOContainer.getClazz());
+        assertTrue(dto.equals(dto));
+        assertFalse(dto.equals(null));
+        assertFalse(dto.equals(new Object()));
+    }
+
+    @Test
+    public void testSerialize() throws ReflectiveOperationException, JsonProcessingException, URISyntaxException {
+        final Object dto = newDTOInstance(DTOContainer.getClazz());
+        final String json = objectMapper.writeValueAsString(dto);
+        assertTrue(json.startsWith("{"));
+        assertTrue(json.endsWith("}"));
+    }
+
+    @Test
+    public void testDeserialize() throws ReflectiveOperationException, IOException, URISyntaxException {
+        final Object dto = newDTOInstance(DTOContainer.getClazz());
+        final String json = objectMapper.writeValueAsString(dto);
+
+        final ObjectMapper alternativeObjectMapper = new ObjectMapper();
+        alternativeObjectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
```

Comment:
This sounds like we're allowing unknown properties in the json, when we want to deserialize it.
Do we also allow this in production?